src/rust/engine/src/externs.rs

```diff
@@ -500,3 +487,65 @@ where
   mem::forget(cs);
   output
 }
+
+
+
+
+#[repr(u8)]
+enum PythonLogLevel {
+  Debug = 0,
+  Info = 1,
+  Warn = 2,
+  Critical = 3,
+}
+
+///
+/// FfiLogger is an implementation of log::Log which asks the Python logging system to log via cffi.
+///
+struct FfiLogger {}
+
+impl FfiLogger {
+  // init must only be called once in the lifetime of the program. No other loggers may be init'd.
+  // If either of the above are violated, expect a panic.
+  pub fn init(&'static self) {
+    log::set_logger(self).expect(
+      "Failed to set logger (maybe you tried to call init multiple times?)",
+    );
+
+    // TODO: Set this to whatever the value Python is using is.
+    log::set_max_level(log::LevelFilter::Trace);
+  }
+}
+
+impl log::Log for FfiLogger {
+  fn enabled(&self, _metadata: &log::Metadata) -> bool {
+    // TODO: Inspect the current log setting from the python and actually filter.
```

Comment:
I would expect that the `debug!` macro (which you are using) delegates to this at some level to implement laziness?

But I agree that it's not a high priority yet.

Reply:
`enabled` is actually only used by the `log_enabled!` macro - the individual logging macros don't call this at all, which is why `log` calls `self.enabled` :)